### PR TITLE
VideoCommon pipelines ("Abstract Pipeline")

### PR DESCRIPTION
**Pull request by stenzek**
This PR adds support for "abstract pipelines", graphics configurations which can be drawn from videocommon. ~~Depends on PR #5343 for the state structures.~~

The eventual is to use it for all graphics drawing, replacing backend-specific shaders/programs. Currently, support is only implemented for "utility" pipelines. This is the first stage, next we can replace the various shaders currently in the backend with these pipelines (e.g. texture conversion, EFB copies, post-processing).

- [x] Graphics pipeline support
- [x] Compute pipeline support
- [x] Utility pipeline/shader drawing
- [x] Replace backend raster font with VideoCommon implementation

The next stages, replacing the shaders, will be a separate PR, to make reviewing/testing easier.

The line count looks kinda scary at the moment due to the amount of duplicated functionality. Once we drop the backend-specific code, it will drop significantly. :)

### CODE REVIEW

File: Source/Core/VideoBackends/D3D/DXShader.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+ID3D11ComputeShader* DXShader::GetD3DComputeShader() const
+{
+  _assert_(m_stage == ShaderStage::Compute);
+  return static_cast<ID3D11ComputeShader*>(m_shader);
+}
+
+bool DXShader::HasBinary() const
+{
+  // m_bytecode should always be non-null.
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXShader.cpp]: Should you assert here?  Or possibly return false if m_bytecode is null?

File: Source/Core/VideoBackends/D3D/DXShader.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  BinaryData ret;
+  ret.resize(m_bytecode->Size());
+  std::memcpy(ret.data(), m_bytecode->Data(), ret.size());
+  return ret;
+}
+
+std::unique_ptr<AbstractShader> Renderer::CreateShaderFromSource(ShaderStage stage,
+                                                                 const char* source, size_t length)
+{
+  D3DBlob* bytecode;
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXShader.cpp]: Maybe make this function leverage `CreateProgramFromBinary` or have them both use the same functionality somehow?

File: Source/Core/VideoBackends/Vulkan/VKShader.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+VkShaderModule VKShader::GetShaderModule() const
+{
+  return m_module;
+}
+
+bool VKShader::HasBinary() const
+{
+  // m_spv should always be non-empty.
```
Comment (iwubcode) [inline Source/Core/VideoBackends/Vulkan/VKShader.cpp]: Ditto my assert comment earlier?

File: Source/Core/VideoBackends/Vulkan/VKShader.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  return std::make_unique<VKShader>(stage, std::move(spv), mod);
+}
+
+std::unique_ptr<AbstractShader> Renderer::CreateProgramFromBinary(ShaderStage stage,
+                                                                  const void* data, size_t length)
+{
+  ShaderCompiler::SPIRVCodeVector spv;
+  const size_t size_in_words = sizeof(length) / sizeof(ShaderCompiler::SPIRVCodeType);
+  spv.resize(length / size_in_words);
+  if (size_in_words > 0)
```
Comment (iwubcode) [inline Source/Core/VideoBackends/Vulkan/VKShader.cpp]: Is this needed?  If size-in-words is 0, `spv.resize` above will cause divide by zero error..

File: Source/Core/VideoBackends/D3D/DXPipeline.cpp [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+  return true;
+}
+
+void Renderer::DestroyUtilityResources()
+{
+  SAFE_RELEASE(m_utility_vertex_buffer);
+  SAFE_RELEASE(m_utility_uniform_buffer);
+}
+
+void Renderer::DrawUtilityPipeline(const AbstractGraphicsPipeline* pipeline,
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXPipeline.cpp]: Is there a reason all this `Renderer` stuff lives in the pipeline class?  I might be in the minority but I'm not  usually a fan of intermingling classes unless there's a good reason.

File: Source/Core/VideoBackends/D3D/DXPipeline.cpp [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+#include "VideoBackends/D3D/DXPipeline.h"
+#include "VideoBackends/D3D/DXShader.h"
+#include "VideoBackends/D3D/DXTexture.h"
+#include "VideoBackends/D3D/FramebufferManager.h"
+#include "VideoBackends/D3D/Render.h"
+#include "VideoBackends/D3D/VertexManager.h"
+
+namespace DX11
+{
+constexpr u32 UTILITY_VBO_SIZE = 512 * 1024;
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXPipeline.cpp]: Just curious, any reason you picked these numbers?

File: Source/Core/VideoBackends/D3D/DXPipeline.cpp [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    m_rasterizer_state->Release();
+  if (m_depth_state)
+    m_depth_state->Release();
+  if (m_blend_state)
+    m_blend_state->Release();
+}
+
+bool DXGraphicsPipeline::HasGeometryShader() const
+{
+  return (m_geometry_shader != nullptr);
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXPipeline.cpp]: Do you need parenthesis?

File: Source/Core/VideoCommon/RenderBase.h [hunk 8]
```diff
@@ -141,7 +161,25 @@ class Renderer
   virtual void ChangeSurface(void* new_surface_handle) {}
   bool UseVertexDepthRange() const;
 
+  // Drawing utility shaders.
+  virtual void DrawUtilityPipeline(const AbstractGraphicsPipeline* pipeline,
```
Comment (lioncash) [inline Source/Core/VideoCommon/RenderBase.h]: Is it possible these functions are doing too much or abstractions aren't done as well they could be? The parameter lists for both of these are kind of ridiculously large.

File: Source/Core/VideoCommon/TextureConfig.h [hunk 9]
```diff
@@ -13,15 +13,24 @@
...
   DXT5,
-  BPTC
+  BPTC,
+  Undefined
 };
 
 struct TextureConfig
 {
   constexpr TextureConfig() = default;
+  TextureConfig(u32 width_, u32 height_, u32 levels_, u32 layers_, AbstractTextureFormat format_,
```
Comment (lioncash) [inline Source/Core/VideoCommon/TextureConfig.h]: As far as I can tell, this can also be constexpr.

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 10]
```diff
@@ -258,6 +271,77 @@ bool Renderer::CheckForHostConfigChanges()
...
+  // Create pipeline.
+  AbstractGraphicsPipelineConfig pconfig = {};
+  pconfig.vertex_format = m_raster_font_vertex_format.get();
+  pconfig.vertex_shader = vs.get();
+  pconfig.pixel_shader = ps.get();
+  pconfig.rasterization_state.hex = RenderState::GetNoCullRasterizationState().hex;
+  pconfig.rasterization_state.primitive = PrimitiveType::Triangles;
+  pconfig.depth_state.hex = RenderState::GetNoDepthTestingDepthStencilState().hex;
+  pconfig.blending_state.hex = RenderState::GetNoBlendingBlendState().hex;
+  pconfig.blending_state.blendenable = true;
```
Comment (iwubcode) [inline Source/Core/VideoCommon/RenderBase.cpp]: Potentially silly question: are these individual values set necessary with the `RenderState::GetNoBlendingBlendState()` call above?

**Comment by bb010g** [comment]
Is this on hold for now?

**Comment by stenzek** [comment]
@bb010g For now. I'm busy IRL for a few weeks, but I'll come back to it at some point.

**Comment by Bombastisch** [comment]
I hope I don't annoy somebody with this question, but is this still on hold? I am always very excited about your PRs.

**Comment by Bombastisch** [comment]
I am sorry to slide in here as a generic user, but since this project was on hold for very long and now the WIP ticket was removed, what is the status of this PR?

**Comment by Bombastisch** [comment]
In this (meanwhile old) article,
https://de.dolphin-emu.org/blog/2013/10/12/d3d9-why-its-not-part-dolphins-future/

It said at the that these compute shaders will give a nice speedboost (it's written in the end of the text). Will we still benefit of that or got Dolphin so fast by now that we won't gain more speed by this?

**Comment by degasus** [comment]
@Bombastisch Please don't abuse github to spam. Use the forum instead. For your question: https://dolphin-emu.org/blog/2017/04/05/dolphin-progress-report-march-2017/#50-3220-gpu-texture-decoding-by-stenzek

**Comment by nbohr1more** [comment]
Not to stir the kettle but I never saw any rebase on this prior to merging.... ?

**Comment by JosJuice** [comment]
This PR was rebased before being merged, not that that actually is a requirement. What we do require is that any small fixup commits are squashed away, but there aren't any commits like that in this PR. Making multiple commits is fine if there's a reason for it.

**Comment by nbohr1more** [comment]
I confirmed that this broke FXAA shading in Vulkan (probably other shader post effects). I was hoping it was an easier answer like rebasing or that some other smaller commit was the cause... :(

**Comment by stenzek** [comment]
@nbohr1more thanks for the report, see #6414.

**Pull request merged**